File: problems/halim/permutations.cpp
#include "header.h"
using namespace std;

vector<string> permutations(const string &s) {
    vector<string> res;
    string path;
    vector<bool> used(s.size(), false);

    function<void(int)> dfs = [&](int start) {
        if (start == (int)s.size()) {
            res.push_back(path);
            return;
        }

        for (int i = 0; i < (int)s.size(); i++) {
            if (used[i]) continue;

            path.push_back(s[i]);
            used[i] = true;
            dfs(start + 1);

            path.pop_back();
            used[i] = false;
        }
    };
    dfs(0);
    return res;
}

int main() {
    string s = "ABCDEFGHIJ";
    vector<string> res = permutations(s);

    ofstream fout("permutations.txt");
    for (auto &p : res) {
        fout << p << "\n";
    }
    fout.close();
    return 0;
}

File: problems/halim/permutations.py
from typing import List
import os


def permutations(s: str) -> List[List[str]]:
    res = []
    path = []
    used = [False] * len(s)

    def dfs(j, path):
        if j == len(s):
            res.append(path.copy())
            return

        for i, c in enumerate(s):
            if used[i]:
                continue

            path.append(c)
            used[i] = True
            dfs(j+1, path)

            path.pop()
            used[i] = False

    dfs(0, [])
    return res


def main():
    s = "ABCDEFGHIJ"
    res = permutations(s)

    script_dir = os.path.dirname(os.path.abspath(__file__))
    output_path = os.path.join(script_dir, "permutations.txt")
    with open(output_path, "w") as f:
        for p in res:
            f.write("".join(p) + "\n")


if __name__ == "__main__":
    main()
